kernel/src/loader/mod.rs

```diff
@@ -144,6 +147,86 @@ where
     Ok(GuestAddress(ehdr.e_entry as usize))
 }
 
+#[cfg(target_arch = "aarch64")]
+pub fn load_kernel<F>(
+    guest_mem: &GuestMemory,
+    kernel_image: &mut F,
+    start_address: usize,
+) -> Result<GuestAddress>
+where
+    F: Read + Seek,
+{
+    /* Kernel boot protocol is specified in the kernel docs
+    Documentation/arm/Booting and Documentation/arm64/booting.txt.
+
+    ======aarch64 kernel header========
+    u32 code0;			/* Executable code */
+    u32 code1;			/* Executable code */
+    u64 text_offset;		/* Image load offset, little endian */
+    u64 image_size;		/* Effective Image size, little endian */
+    u64 flags;			/* kernel flags, little endian */
+    u64 res2	= 0;		/* reserved */
+    u64 res3	= 0;		/* reserved */
+    u64 res4	= 0;		/* reserved */
+    u32 magic	= 0x644d5241;	/* Magic number, little endian, "ARM\x64" */
+    u32 res5;			/* reserved (used for PE COFF offset) */
+    ====================================
+     */
+    const AARCH64_KERNEL_LOAD_ADDR: usize = 0x80000;
+    const AARCH64_MAGIC_NUMBER: u32 = 0x644d5241;
+    let mut kernel_load_offset = AARCH64_KERNEL_LOAD_ADDR;
+
+    /* Look for the magic number inside the elf header. */
+    let magic_offset_header = 2 * mem::size_of::<u32>() + 6 * mem::size_of::<u64>(); // This should total 56.
+
+    kernel_image
+        .seek(SeekFrom::Start(magic_offset_header as u64))
+        .map_err(|_| Error::SeekKernelImage)?;
+    let mut magic_number: u32 = 0;
+    unsafe {
+        sys_util::read_struct(kernel_image, &mut magic_number)
+            .map_err(|_| Error::ReadProgramHeader)?
+    }
+    if magic_number != AARCH64_MAGIC_NUMBER {
+        return Err(Error::InvalidElfMagicNumber);
+    }
+
+    /* Look for the `text_offset` from the elf header. */
+    let text_offset = 2 * mem::size_of::<u32>() as u64;
+    kernel_image
+        .seek(SeekFrom::Start(text_offset)) // This should total 8.
+        .map_err(|_| Error::SeekKernelImage)?;
+    let mut hdrvals: [u64; 2] = [0; 2];
```

Comment:
What's the endianness of `aarch64`? If it's not always little endian (I've heard ARM can be either big or little, but don't know much more about that), then we must make sure the values of `text_offset`, `image_size`, `flags` and `magic` are properly read.

They're always written using little endian in the file, so we'd have to add something like `magic_number = magic_number.from_le();` after reading the value from the file. This is a no-op on little endian, and does the proper conversion on big endian systems (same for `hdrvals[0]`, etc.).

```

Reply:
Added from_le for the values read from the header, Take a look.